apps/forklift/lib/forklift/data_writer.ex

```diff
@@ -189,9 +164,11 @@ defmodule Forklift.DataWriter do
     datum |> Forklift.Util.add_to_metadata(:forklift_start_time, Data.Timing.current_time())
   end
 
-  defp add_ingestion_info(data) do
-    payload = Map.get(data, :payload, %{})
+  defp add_ingestion_info(%{payload: payload} = data) when not is_map(payload) do
```

Comment:
Hmm, should we be more specific here? Besides end_of_data, if the payload isn't a map, I would want an error to be raised. What is payload when it's an end_of_data message? Is it just a string that we could match on exactly in the hydra?